scripts/test.py

```diff
@@ -0,0 +1,146 @@
+#!/usr/bin/env python3
+
+# Copyright (c) 2019 Advanced Micro Devices, Inc. All rights reserved
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at:
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+'''
+GFX reconstruct build script
+'''
+
+import argparse
+import copy
+import distutils.version
+import importlib
+import os
+import platform
+import re
+import shutil
+import subprocess
+import sys
+
+
+def is_windows():
+    '''
+    Check if the system is Windows
+    '''
+    return 'windows' == platform.system().lower()
+
+
+BUILD_ROOT = os.path.abspath(os.path.join(
+    os.path.split(os.path.abspath(__file__))[0], '..'))
+ALL_TESTS = {
+    'gfxrecon_application_test': [],
+    'gfxrecon_decode_test': [],
+    'gfxrecon_encode_test': [],
+    'gfxrecon_format_test': [],
+    'gfxrecon_util_test': [],
+    'VkLayer_gfxreconstruct_test': [],
+}
+
+
+class TestError(Exception):
+    '''
+    Raised on a test error
+    '''
+
+
+def import_build_script():
+    '''
+    Import the build script - reuse build script constants
+    '''
+    sys.path.insert(0, os.path.join(BUILD_ROOT, 'scripts'))
+    build_script = importlib.import_module('build')
+    return build_script
+
+
+def parse_args(build_script):
+    '''
+    Parse command line arguments
+    '''
+    arg_parser = argparse.ArgumentParser(
+        description="gfxreconstruct build script",
+        formatter_class=argparse.ArgumentDefaultsHelpFormatter)
+    arg_parser.add_argument('--version', dest='version',
+                            action='version', version=str(build_script.VERSION))
+    arg_parser.add_argument(
+        '--test-exe', dest='test_exe', metavar='TEST_EXE',
+        action='store', default=None,
+        help='Test executable to run')
```

Comment:
There's only a small number of valid test executables.  Add a `choices=` parameter here to ensure that, if `test_exe` is set, it's valid.

Reply:
There are a few of problems with adding choices
1. The test exe includes the local system path, which cannot easily be added to the choice
2. It limits the user from adding tests which are not built as part of GFXReconstruct - we have such tests intenrally
3. It makes this argument required. The current behavior is that if this argument is omitted, all tests are executed - this can fixed by adding an 'all' test exe, but I find this to be ugly.

Reply:
Ah, all right.  There's confusion here because the code implies that all the tests are listed in `ALL_TESTS`.  Please add script comments and help text to make it clear that `--test-exe` is not required to be a member of `ALL_TESTS`.  I think it would be helpful to rename `ALL_TESTS` to `ALL_INTERNAL_TESTS` to help make it clear that external tests are also supported.

I'd like the help text to be much more complete in this case, e.g.:
```
help='the name of a single internal test to run (one of: {}), or a path '
     'to an external test to run.  If not specified, all internal tests are run.'
     .format(', '.join(ALL_INTERNAL_TESTS))